NExtends/Expressions/ExpressionHelper.cs

```diff
@@ -0,0 +1,238 @@
+		{
+			var visitor = new HasTypeVisitor<TTest>();
+			return visitor.VisitRoot(original);
+		}
+
+		private class HasTypeVisitor<TExpression> : ExpressionVisitor
+		{
+			private Type _toKeep;
+			private bool _HasType = false;
+
+			public HasTypeVisitor()
+			{
+				_toKeep = typeof(TExpression);
+			}
+
+			public bool VisitRoot(Expression node)
+			{
+				var visited = Visit(node);
+				return _HasType;
+			}
+
+			public override Expression Visit(Expression node)
+			{
+				if (_HasType) { return node; }
+				if (node != null && node.Type == _toKeep)
+				{
+					_HasType = true;
+					return node;
+				}
+				else
+				{
+					return base.Visit(node);
+				}
+			}
+		}
+
+		public ICollection<Expression<Func<TEntity, object>>> Remove<TEntity>(IEnumerable<Expression<Func<TEntity, object>>> source, Expression<Func<TEntity, object>> candidate)
+		{
+			var currentIncludes = source;
+
+			var result = new HashSet<Expression<Func<TEntity, object>>>();
+
+			foreach (var exp in source)
+			{
+				if (!Equals(exp, candidate))
+				{
+					result.Add(exp);
+				}
+			}
+
+			return result;
+		}
+
+		public bool Equals<TEntity>(Expression<Func<TEntity, object>> exp1, Expression<Func<TEntity, object>> exp2)
+		{
+			return GetPropertyFromPropertySelector<TEntity>(exp1) == GetPropertyFromPropertySelector<TEntity>(exp2);
+		}
+
+		/// <summary>
+		/// Récupère la propriété ciblée par une expression, notamment pour voir si 2 expressions sont égales = récupèrent la même propriété
+		/// </summary>
+		/// <typeparam name="TEntity"></typeparam>
+		/// <param name="propertySelector"></param>
+		/// <returns></returns>
+		private PropertyInfo GetPropertyFromPropertySelector<TEntity>(Expression<Func<TEntity, object>> propertySelector)
+		{
+			var propertyInfo = (propertySelector.Body as MemberExpression).Member as PropertyInfo;
+			if (propertyInfo == null)
+			{
+				throw new ArgumentException("The lambda expression 'property' should point to a valid Property");
+			}
+			return propertyInfo;
+		}
+
+		public Expression NestedPropertyAccessor(ParameterExpression seed, string field)
+		{
+			PropertyInfo unusedFinalProperty;
+			return NestedPropertyAccessor(seed, field, out unusedFinalProperty);
+		}
+		public Expression NestedPropertyAccessor(ParameterExpression seed, string field, out PropertyInfo property)
+		{
+			var type = seed.Type;
+			return NestedPropertyAccessor(type, seed, field, out property);
+		}
+		public Expression NestedPropertyAccessor(Type type, ParameterExpression seed, string field, out PropertyInfo property)
+		{
+			return NestedPropertyAccessor(type, seed, field.Split('.'), out property);
+		}
+		private Expression NestedPropertyAccessor(Type type, ParameterExpression seed, string[] fields, out PropertyInfo property)
+		{
+			property = null;
+			Expression body = seed;
+
+			foreach (var member in fields)
+			{
+				property = type.GetProperties().Where(p => p.Name.ToLower() == member.ToLower()).FirstOrDefault();
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Drop 'Where' and move the condition into the 'FirstOrDefault'. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS2971)